server/src/main/java/org/elasticsearch/index/search/MatchQuery.java

```diff
@@ -574,30 +581,37 @@ private Query analyzeTerm(String field, TokenStream stream, boolean isPrefix) th
             return isPrefix && lastOffset == offsetAtt.endOffset() ? newPrefixQuery(term) : newTermQuery(term);
         }
 
-        private void add(BooleanQuery.Builder q, String field, List<Term> current, BooleanClause.Occur operator, boolean isPrefix) {
-            if (current.isEmpty()) {
-                return;
+        @Override
+        protected Query analyzeBoolean(String field, TokenStream stream) throws IOException {
+            TermToBytesRefAttribute termAtt = stream.getAttribute(TermToBytesRefAttribute.class);
+            TypeAttribute typeAtt = stream.getAttribute(TypeAttribute.class);
+
+            stream.reset();
+            List<TermAndBoost> terms = new ArrayList<>();
+            while (stream.incrementToken()) {
+                terms.add(new TermAndBoost(new Term(field, termAtt.getBytesRef()), getTokenTypeBoost(typeAtt)));
             }
-            if (current.size() == 1) {
-                if (isPrefix) {
-                    q.add(newPrefixQuery(current.get(0)), operator);
-                } else {
-                    q.add(newTermQuery(current.get(0)), operator);
-                }
-            } else {
-                // We don't apply prefix on synonyms
-                q.add(newSynonymQuery(current.toArray(new Term[current.size()])), operator);
+            return newSynonymQuery(field, terms);
+        }
+
+        // TODO maybe better to add newSynonymQuery() variant to org.apache.lucene.util.QueryBuilder that also takes boost arguments?
```

Comment:
@jimczi do you agree that it would make sense to try and push this method to QueryBuilder? I can open a PR in that case, but we can keep this here for the time being.